### PR TITLE
fix(fslib): don't close zip files with open handles in ZipOpenFS

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

- When opening files inside zip archives with ZipOpenFS, with `maxOpenFiles` set, it will eventually close the zip archives even though they have open file handles. This causes a `EBUSY: archive closed` exception when trying to use the file handle.
- `getZip*` calls `limitOpenFiles` after adding the ZipFS entry to the cache which can cause it to immediately close the file it just opened causing a `EBUSY: archive closed` exception.
Only way to have this happen before this PR is to set `maxOpenFiles: 0`

Found these issues when looking into https://github.com/yarnpkg/berry/issues/1576

**How did you fix it?**

- Made `limitOpenFiles` skip ZipFS instances with open file handles.
- Made `getZipSync` run `limitOpenFiles` before adding the file back into the cache.

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I have verified that all automated PR checks pass.

### CODE REVIEW

File: packages/yarnpkg-fslib/sources/ZipFS.ts [hunk 1]
```diff
@@ -485,16 +489,21 @@ export class ZipFS extends BasePortableFakeFS {
       },
     });
 
+    const fd = this.openSync(p, ``);
```
Comment (arcanis) [inline packages/yarnpkg-fslib/sources/ZipFS.ts:492]: Why an empty string? Why not `r`?

File: packages/yarnpkg-fslib/sources/ZipFS.ts [hunk 2]
```diff
@@ -485,16 +489,21 @@ export class ZipFS extends BasePortableFakeFS {
...
+
     const immediate = setImmediate(() => {
       try {
         const data = this.readFileSync(p, encoding);
 
         stream.bytesRead = data.length;
-        stream.write(data);
-        stream.end();
+        stream.end(data);
+        stream.destroy();
```
Comment (arcanis) [inline packages/yarnpkg-fslib/sources/ZipFS.ts:500]: Isn't `destroy` automatically closed on `end`?

Comment (merceyz) [reply packages/yarnpkg-fslib/sources/ZipFS.ts:500]: Interestingly, no. The `close` event is only called when using `destroy`

File: packages/yarnpkg-fslib/sources/ZipFS.ts (hunk 1 above)
Comment (merceyz) [reply packages/yarnpkg-fslib/sources/ZipFS.ts]: The flag isn't used for anything so I left it out, but meant to fix it... then I forgot to fix it.

**Pull request merged**